clippy_lints/src/methods/string_as_str.rs

```diff
@@ -0,0 +1,60 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::is_type_diagnostic_item};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::LateContext;
+
+use super::STRING_AS_STR;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>) {
+    if let ExprKind::MethodCall(path, sub_expr, ..) = &expr.kind {
+        if is_string(cx, sub_expr) && !is_within_borrowed_call(cx, sub_expr) {
+            let var_name = if let ExprKind::Path(QPath::Resolved(_, path)) = sub_expr.kind {
+                path.segments.first().map(|name| name.ident.name.as_str())
+            } else {
+                None
+            };
+
+            span_lint_and_sugg(
+                cx,
+                STRING_AS_STR,
+                expr.span,
+                &format!("used `{}` for reference", path.ident.name),
+                "replace it with",
+                format!("&{}", var_name.unwrap_or("")),
+                Applicability::MaybeIncorrect,
+            );
+        }
+    }
+}
+
+fn is_string(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    is_type_diagnostic_item(
+        cx,
+        cx.typeck_results().expr_ty(expr).peel_refs(),
```

Comment:
```suggestion
        cx.typeck_results().expr_ty(expr),
```

I think this should just be this and with that you probably won't need the `is_within_borrowed_call` function. The `peel_refs` causes the lint to trigger on `&String` types.